config: Add CMake printf variable and C library variables to the generated CMake module

hugueskamba opened the pull request:
`MBED_TARGET_SUPPORTED_C_LIBS`, `MBED_PRINTF_LIB` and `MBED_C_LIB` are provided
in the generated configuration CMake module to allow the toolchain to be
configured for the selected printf library and C library.

### Description

<!--
Please add any detail or context that would be useful to a reviewer.
-->



### Test Coverage

<!--
Please put an `x` in the correct box e.g. `[x]` to indicate the testing coverage of this change.
-->

- [X]  This change is covered by existing or additional automated tests.
- [ ]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

Patater commented:
This is failing the coding style check. This change also needs a news file.

hugueskamba commented:
> This is failing the coding style check. This change also needs a news file.

Where can I see the coding style failures to see what it expects?
What is a "news file"? Where can I see examples of it in the repo so a new one can be added?

rwalton-arm commented:
> > This is failing the coding style check. This change also needs a news file.
> 
> Where can I see the coding style failures to see what it expects? When I click on ["Check failure on line 36 in Build log"](https://github.com/ARMmbed/mbed-tools/pull/73/files#annotation_433858682) it takes me to the "Files changed" tab with nothing pointing to the code style error.
> 
> What is a "news file"? Where can I see examples of it in the repo so a new one can be added?

News files explained here: https://github.com/ARMmbed/mbed-tools/blob/master/CONTRIBUTING.md
Code style checks explained here: https://github.com/ARMmbed/mbed-tools/blob/master/DEVELOPMENT.md

hugueskamba commented:
> > > This is failing the coding style check. This change also needs a news file.
> > 
> > 
> > Where can I see the coding style failures to see what it expects? When I click on ["Check failure on line 36 in Build log"](https://github.com/ARMmbed/mbed-tools/pull/73/files#annotation_433858682) it takes me to the "Files changed" tab with nothing pointing to the code style error.
> > What is a "news file"? Where can I see examples of it in the repo so a new one can be added?
> 
> News files explained here: https://github.com/ARMmbed/mbed-tools/blob/master/CONTRIBUTING.md
> Some examples of news files are here: https://github.com/ARMmbed/mbed-tools/tree/master/news
> Code style checks explained here: https://github.com/ARMmbed/mbed-tools/blob/master/DEVELOPMENT.md

I have added a news file and corrected the coding style issue using the `pre-commit` command.
Thanks.

rwalton-arm commented on tests/build/_internal/test_cmake_file.py:
```diff
@@ -40,13 +43,19 @@ def test_returns_rendered_content(self):
         for macro in target["features"] + target["components"] + [toolchain_name]:
             self.assertIn(macro, result)
+
+        for supported_c_lib in target["supported_c_libs"]:
```
Looks like we want to iterate over the `.values()` here, at the moment we're checking the key (which is the toolchain name) is in the rendered template.

hugueskamba replied:
Is [this](https://github.com/ARMmbed/mbed-tools/compare/b0e0c663fb9e74f4ce5d2e2d1c29e311d77c594f..2797d81a6358ac8c7bd4eb63918f31fdb670c305) ok?

rwalton-arm replied:
Looks ok, but you've force pushed over my commit to pin the `black` version. Please reinstate `black = "==19.10b0"`.

hugueskamba replied:
I restored it right after. See the current state of that file in https://github.com/ARMmbed/mbed-tools/pull/73/files

rwalton-arm replied:
Sorry. Missed that. LGTM.

rwalton-arm commented on news/20200915.feature:
```diff
@@ -0,0 +1 @@
+Add CMake variables for the selected printf and C library to the generated CMake module
```
Add a newline to the end of this file please.

hugueskamba replied:
Is [this](https://github.com/ARMmbed/mbed-tools/compare/b0e0c663fb9e74f4ce5d2e2d1c29e311d77c594f..2797d81a6358ac8c7bd4eb63918f31fdb670c305) ok?

rwalton-arm replied:
Looks good. Thanks.

rwalton-arm commented on src/mbed_tools/build/_internal/templates/mbed_config.tmpl:
```diff
@@ -7,6 +7,14 @@
+set(MBED_C_LIB "{{c_lib}}" CACHE STRING "")
+set(MBED_PRINTF_LIB "{{printf_lib}}" CACHE STRING "")
+
+list(APPEND MBED_TARGET_SUPPORTED_C_LIBS {% if toolchain_name.lower() in supported_c_libs -%} {% for supported_c_lib in supported_c_libs[toolchain_name.lower()] %}
```
Should we move the logic to get the list of supported c libs into the python module?

hugueskamba replied:
Is [this](https://github.com/ARMmbed/mbed-tools/compare/cd1d24ce4340bf582e545f3c869b1f0e87a2657b..aeec090577452b406c64135aff8a1db84904ec36) ok?

rwalton-arm replied:
Much better. Thanks.
